File: app/src/main/java/com/canli/oya/traininventory/adapters/CategoryAdapter.java
package com.canli.oya.traininventory.adapters;

import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.canli.oya.traininventory.BR;
import com.canli.oya.traininventory.R;
import com.canli.oya.traininventory.databinding.CategoryItemBinding;

import java.util.List;

public class CategoryAdapter extends RecyclerView.Adapter<CategoryAdapter.CategoryHolder> {

    private List<String> mCategoryList;
    private CategoryItemClickListener mClickListener;

    public CategoryAdapter(CategoryItemClickListener listener) {
        mClickListener = listener;
    }

    @NonNull
    @Override
    public CategoryHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        CategoryItemBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.category_item,
                        parent, false);
        binding.setCategoryItemClick(mClickListener);
        return new CategoryHolder(binding);
    }

    public void setCategories(List<String> newList) {
        mCategoryList = newList;
        notifyDataSetChanged();
    }

    @Override
    public void onBindViewHolder(@NonNull CategoryHolder holder, int position) {
        String currentCategory = mCategoryList.get(position);
        holder.binding.setVariable(BR.categoryName, currentCategory);
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mCategoryList == null ? 0 : mCategoryList.size();
    }

    class CategoryHolder extends RecyclerView.ViewHolder {
        final CategoryItemBinding binding;

        CategoryHolder(CategoryItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    public interface CategoryItemClickListener {
        void onCategoryItemClicked(String categoryName);
    }
}

File: app/src/main/java/com/canli/oya/traininventoryroom/adapters/CategoryAdapter.kt
package com.canli.oya.traininventoryroom.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.canli.oya.traininventoryroom.R
import com.canli.oya.traininventoryroom.databinding.ItemCategoryBinding

class CategoryAdapter (private val mClickListener: CategoryItemClickListener) : RecyclerView.Adapter<CategoryAdapter.CategoryHolder>() {

    var categoryList: List<String>? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryHolder {
        val binding = DataBindingUtil
                .inflate<ItemCategoryBinding>(LayoutInflater.from(parent.context), R.layout.item_category,
                        parent, false)
        binding.categoryItemClick = mClickListener
        return CategoryHolder(binding)
    }

    override fun onBindViewHolder(holder: CategoryHolder, position: Int) {
        val currentCategory = categoryList?.get(position)
        holder.binding.categoryName = currentCategory
        holder.binding.executePendingBindings()
        holder.binding.categoryItemNumber.text = "${position+1}."
    }

    override fun getItemCount() = categoryList?.size ?: 0

    inner class CategoryHolder(val binding: ItemCategoryBinding) : RecyclerView.ViewHolder(binding.root)

    interface CategoryItemClickListener {
        fun onCategoryItemClicked(categoryName: String)
    }
}
